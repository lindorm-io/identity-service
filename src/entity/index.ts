export * from "./DisplayName";
export * from "./Identity";
