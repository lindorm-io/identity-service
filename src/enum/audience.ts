export enum Audience {
  ACCESS = "access",
}
