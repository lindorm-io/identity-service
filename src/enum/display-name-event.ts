export enum DisplayNameEvent {
  CREATED = "display_name_created",

  NUMBER_ADDED = "display_name_number_added",
  NUMBER_REMOVED = "display_name_number_removed",
}
