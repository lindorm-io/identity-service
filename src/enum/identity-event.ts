export enum IdentityEvent {
  CREATED = "identity_created",

  ADDRESS_CHANGED = "identity_address_changed",
  BIRTH_DATE_CHANGED = "identity_birth_date_changed",
  DISPLAY_NAME_CHANGED = "identity_display_name_changed",
  FAMILY_NAME_CHANGED = "identity_family_name_changed",
  GENDER_CHANGED = "identity_gender_changed",
  GIVEN_NAME_CHANGED = "identity_given_name_changed",
  GRAVATAR_CHANGED = "identity_gravatar_changed",
  LOCALE_CHANGED = "identity_locale_changed",
  MIDDLE_NAME_CHANGED = "identity_middle_name_changed",
  NICKNAME_CHANGED = "identity_nickname_changed",
  PHONE_NUMBER_CHANGED = "identity_phone_number_changed",
  PHONE_NUMBER_VERIFIED_CHANGED = "identity_phone_number_verified_changed",
  PICTURE_CHANGED = "identity_picture_changed",
  PREFERRED_USERNAME_CHANGED = "identity_preferred_username_changed",
  PROFILE_CHANGED = "identity_profile_changed",
  USERNAME_CHANGED = "identity_username_changed",
  WEBSITE_CHANGED = "identity_website_changed",
  ZONE_INFO_CHANGED = "identity_zone_info_changed",
}
