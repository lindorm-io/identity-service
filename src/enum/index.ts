export * from "./display-name-event";
export * from "./identity-event";
