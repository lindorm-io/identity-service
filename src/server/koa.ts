import { KoaApp } from "@lindorm-io/koa";
import { SERVER_PORT, TOKEN_ISSUER_MW_OPTIONS } from "../config";
import { appRoot, headless, identity } from "../route";
import { getMongoMiddleware, getWebKeyMiddleware, repositoryMiddleware } from "../middleware";
import { tokenIssuerMiddleware } from "@lindorm-io/koa-jwt";
import { winston } from "../logger";

export const koa = new KoaApp({
  logger: winston,
  port: SERVER_PORT,
});

koa.addMiddleware(getMongoMiddleware());
koa.addMiddleware(repositoryMiddleware);
koa.addMiddleware(getWebKeyMiddleware());
koa.addMiddleware(tokenIssuerMiddleware(TOKEN_ISSUER_MW_OPTIONS));

koa.addRoute("/", appRoot);
koa.addRoute("/headless", headless);
koa.addRoute("/identity", identity);
