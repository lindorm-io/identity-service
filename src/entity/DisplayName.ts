import { DisplayNameEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { includes } from "lodash";

export interface IDisplayName extends IEntity {
  name: string;
  numbers: Array<number>;
}

export interface IDisplayNameOptions extends IEntityBaseOptions {
  name: string;
  numbers?: Array<number>;
}

export class DisplayName extends EntityBase {
  private _name: string;
  private _numbers: Array<number>;

  constructor(options: IDisplayNameOptions) {
    super(options);
    this._name = options.name;
    this._numbers = options.numbers || [];
  }

  public get name(): string {
    return this._name;
  }

  public get numbers(): Array<number> {
    return this._numbers;
  }

  public add(number: number): void {
    if (includes(this._numbers, number)) {
      throw new Error("Number already exists for this DisplayName");
    }

    this._numbers.push(number);
    this.addEvent(DisplayNameEvent.NUMBER_ADDED, { number });
  }

  public remove(number: number): void {
    const array: Array<number> = [];

    for (const item of this._numbers) {
      if (item === number) continue;
      array.push(item);
    }

    this._numbers = array;
    this.addEvent(DisplayNameEvent.NUMBER_REMOVED, { number });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DisplayNameEvent.CREATED) continue;
      throw new EntityCreationError("DisplayName");
    }

    this.addEvent(DisplayNameEvent.CREATED, {
      name: this._name,
      numbers: this._numbers,
    });
  }
}
