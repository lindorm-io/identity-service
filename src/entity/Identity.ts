import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { IdentityEvent } from "../enum";

export interface IIdentityAddress {
  country: string;
  locality: string;
  postalCode: string;
  region: string;
  streetAddress: string;
}

export interface IIdentityDisplayName {
  name: string;
  number: number;
}

export interface IIdentity extends IEntity {
  // public
  displayName: IIdentityDisplayName;
  gravatar: string;

  // private
  username: string;

  // openid
  address: IIdentityAddress;
  birthDate: string;
  familyName: string;
  gender: string;
  givenName: string;
  locale: string;
  middleName: string;
  nickname: string;
  phoneNumber: string;
  phoneNumberVerified: boolean;
  picture: string;
  preferredUsername: string;
  profile: string;
  website: string;
  zoneInfo: string;
}

export interface IIdentityOptions extends IEntityBaseOptions {
  // public
  displayName?: IIdentityDisplayName;
  gravatar?: string;

  // private
  username?: string;

  // openid
  address?: IIdentityAddress;
  birthDate?: string;
  familyName?: string;
  gender?: string;
  givenName?: string;
  locale?: string;
  middleName?: string;
  nickname?: string;
  phoneNumber?: string;
  phoneNumberVerified?: boolean;
  picture?: string;
  preferredUsername?: string;
  profile?: string;
  website?: string;
  zoneInfo?: string;
}

export class Identity extends EntityBase {
  private _address: IIdentityAddress;
  private _birthDate: string;
  private _displayName: IIdentityDisplayName;
  private _familyName: string;
  private _gender: string;
  private _givenName: string;
  private _gravatar: string;
  private _locale: string;
  private _middleName: string;
  private _nickname: string;
  private _phoneNumber: string;
  private _phoneNumberVerified: boolean;
  private _picture: string;
  private _preferredUsername: string;
  private _profile: string;
  private _username: string;
  private _website: string;
  private _zoneInfo: string;

  constructor(options: IIdentityOptions) {
    super(options);
    this._address = options.address || null;
    this._birthDate = options.birthDate || null;
    this._displayName = options.displayName || null;
    this._familyName = options.familyName || null;
    this._gender = options.gender || null;
    this._givenName = options.givenName || null;
    this._gravatar = options.gravatar || null;
    this._locale = options.locale || null;
    this._middleName = options.middleName || null;
    this._nickname = options.nickname || null;
    this._phoneNumber = options.phoneNumber || null;
    this._phoneNumberVerified = options.phoneNumberVerified || false;
    this._picture = options.picture || null;
    this._preferredUsername = options.preferredUsername || null;
    this._profile = options.profile || null;
    this._username = options.username || null;
    this._website = options.website || null;
    this._zoneInfo = options.zoneInfo || null;
  }

  public get address(): IIdentityAddress {
    return this._address;
  }
  public set address(address: IIdentityAddress) {
    this._address = address;
    this.addEvent(IdentityEvent.ADDRESS_CHANGED, { address: this._address });
  }

  public get birthDate(): string {
    return this._birthDate;
  }
  public set birthDate(birthDate: string) {
    this._birthDate = birthDate;
    this.addEvent(IdentityEvent.BIRTH_DATE_CHANGED, { birthDate: this._birthDate });
  }

  public get displayName(): IIdentityDisplayName {
    return this._displayName;
  }
  public set displayName(displayName: IIdentityDisplayName) {
    this._displayName = displayName;
    this.addEvent(IdentityEvent.DISPLAY_NAME_CHANGED, { displayName: this._displayName });
  }

  public get familyName(): string {
    return this._familyName;
  }
  public set familyName(familyName: string) {
    this._familyName = familyName;
    this.addEvent(IdentityEvent.FAMILY_NAME_CHANGED, { familyName: this._familyName });
  }

  public get gender(): string {
    return this._gender;
  }
  public set gender(gender: string) {
    this._gender = gender;
    this.addEvent(IdentityEvent.GENDER_CHANGED, { gender: this._gender });
  }

  public get givenName(): string {
    return this._givenName;
  }
  public set givenName(givenName: string) {
    this._givenName = givenName;
    this.addEvent(IdentityEvent.GIVEN_NAME_CHANGED, { givenName: this._givenName });
  }

  public get gravatar(): string {
    return this._gravatar;
  }
  public set gravatar(gravatar: string) {
    this._gravatar = gravatar;
    this.addEvent(IdentityEvent.GRAVATAR_CHANGED, { gravatar: this._givenName });
  }

  public get locale(): string {
    return this._locale;
  }
  public set locale(locale: string) {
    this._locale = locale;
    this.addEvent(IdentityEvent.LOCALE_CHANGED, { locale: this._locale });
  }

  public get middleName(): string {
    return this._middleName;
  }
  public set middleName(middleName: string) {
    this._middleName = middleName;
    this.addEvent(IdentityEvent.MIDDLE_NAME_CHANGED, { middleName: this._middleName });
  }

  public get nickname(): string {
    return this._nickname;
  }
  public set nickname(nickname: string) {
    this._nickname = nickname;
    this.addEvent(IdentityEvent.NICKNAME_CHANGED, { nickname: this._nickname });
  }

  public get phoneNumber(): string {
    return this._phoneNumber;
  }
  public set phoneNumber(phoneNumber: string) {
    this._phoneNumber = phoneNumber;
    this.addEvent(IdentityEvent.PHONE_NUMBER_CHANGED, { phoneNumber: this._phoneNumber });
  }

  public get phoneNumberVerified(): boolean {
    return this._phoneNumberVerified;
  }
  public set phoneNumberVerified(phoneNumberVerified: boolean) {
    this._phoneNumberVerified = phoneNumberVerified;
    this.addEvent(IdentityEvent.PHONE_NUMBER_VERIFIED_CHANGED, { phoneNumberVerified: this._phoneNumber });
  }

  public get picture(): string {
    return this._picture;
  }
  public set picture(picture: string) {
    this._picture = picture;
    this.addEvent(IdentityEvent.PICTURE_CHANGED, { picture: this._picture });
  }

  public get preferredUsername(): string {
    return this._preferredUsername;
  }
  public set preferredUsername(preferredUsername: string) {
    this._preferredUsername = preferredUsername;
    this.addEvent(IdentityEvent.PREFERRED_USERNAME_CHANGED, { preferredUsername: this._preferredUsername });
  }

  public get profile(): string {
    return this._profile;
  }
  public set profile(profile: string) {
    this._profile = profile;
    this.addEvent(IdentityEvent.PROFILE_CHANGED, { profile: this._profile });
  }

  public get username(): string {
    return this._username;
  }
  public set username(username: string) {
    this._username = username;
    this.addEvent(IdentityEvent.USERNAME_CHANGED, { username: this._profile });
  }

  public get website(): string {
    return this._website;
  }
  public set website(website: string) {
    this._website = website;
    this.addEvent(IdentityEvent.WEBSITE_CHANGED, { website: this._website });
  }

  public get zoneInfo(): string {
    return this._zoneInfo;
  }
  public set zoneInfo(zoneInfo: string) {
    this._zoneInfo = zoneInfo;
    this.addEvent(IdentityEvent.ZONE_INFO_CHANGED, { zoneInfo: this._zoneInfo });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== IdentityEvent.CREATED) continue;
      throw new EntityCreationError("Identity");
    }

    this.addEvent(IdentityEvent.CREATED, {
      address: this._address,
      birthDate: this._birthDate,
      displayName: this._displayName,
      familyName: this._familyName,
      gender: this._gender,
      givenName: this._givenName,
      gravatar: this._gravatar,
      locale: this._locale,
      middleName: this._middleName,
      nickname: this._nickname,
      phoneNumber: this._phoneNumber,
      phoneNumberVerified: this._phoneNumberVerified,
      picture: this._picture,
      preferredUsername: this._preferredUsername,
      profile: this._profile,
      username: this._username,
      website: this._website,
      zoneInfo: this._zoneInfo,
    });
  }
}
