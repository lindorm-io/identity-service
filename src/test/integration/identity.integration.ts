import MockDate from "mockdate";
import request from "supertest";
import { Audience } from "../../enum";
import { Identity } from "../../entity";
import { Permission, Scope } from "@lindorm-io/jwt";
import { inMemoryStore, setupIntegration, TEST_IDENTITY, TEST_IDENTITY_REPOSITORY, TEST_ISSUER } from "../grey-box";
import { koa } from "../../server/koa";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "e3926ddb-ecaf-4f66-855a-d143af54953c"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("/identity", () => {
  let accessToken: string;

  beforeAll(async () => {
    await setupIntegration();
    koa.load();

    ({ token: accessToken } = TEST_ISSUER.sign({
      audience: Audience.ACCESS,
      expiry: "2 minutes",
      permission: Permission.ADMIN,
      scope: [Scope.DEFAULT, Scope.EDIT, Scope.OPENID].join(" "),
      subject: TEST_IDENTITY.id,
    }));
  });

  test("GET /:id - should return identity information", async () => {
    const response = await request(koa.callback())
      .get(`/identity/${TEST_IDENTITY.id}`)
      .set("Authorization", `Bearer ${accessToken}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .expect(200);

    expect(response.body).toMatchSnapshot();
    expect(inMemoryStore).toMatchSnapshot();
  });

  test("PATCH /:id - should update", async () => {
    const id = "0d573da5-97d7-44b1-8aee-9951e177d0b2";
    await TEST_IDENTITY_REPOSITORY.create(new Identity({ id }));

    await request(koa.callback())
      .patch(`/identity/${id}`)
      .set("Authorization", `Bearer ${accessToken}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .send({
        givenName: "new-givenName",
        middleName: "new-middleName",
      })
      .expect(204);

    await expect(TEST_IDENTITY_REPOSITORY.find({ id })).resolves.toMatchSnapshot();
  });

  test("DELETE /:id - should delete", async () => {
    const id = "91768368-7e70-4cee-a031-deb78e0faebc";
    await TEST_IDENTITY_REPOSITORY.create(new Identity({ id }));

    await request(koa.callback())
      .delete(`/identity/${id}`)
      .set("Authorization", `Bearer ${accessToken}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .expect(202);

    await expect(TEST_IDENTITY_REPOSITORY.find({ id })).rejects.toMatchSnapshot();
  });

  test("PATCH /:id/address", async () => {
    const id = "c8909659-a490-4d6c-88d3-6a01712b7c64";
    await TEST_IDENTITY_REPOSITORY.create(new Identity({ id }));

    await request(koa.callback())
      .patch(`/identity/${id}/address`)
      .set("Authorization", `Bearer ${accessToken}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .send({
        country: "country",
        locality: "locality",
        postalCode: "postalCode",
        region: "region",
        streetAddress: "streetAddress",
      })
      .expect(204);

    await expect(TEST_IDENTITY_REPOSITORY.find({ id })).resolves.toMatchSnapshot();
  });

  test("PATCH /:id/display-name", async () => {
    const id = "b2b2c123-efa5-442a-9ece-5fd81f8c64d9";
    await TEST_IDENTITY_REPOSITORY.create(new Identity({ id }));

    await request(koa.callback())
      .patch(`/identity/${id}/display-name`)
      .set("Authorization", `Bearer ${accessToken}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .send({
        displayName: "displayName",
      })
      .expect(204);

    await expect(TEST_IDENTITY_REPOSITORY.find({ id })).resolves.toStrictEqual(
      expect.objectContaining({
        displayName: { name: "displayName", number: expect.any(Number) },
      }),
    );
  });

  test("PATCH /:id/username", async () => {
    const id = "9318e77b-10bf-413a-89be-ca36056b8723";
    await TEST_IDENTITY_REPOSITORY.create(new Identity({ id }));

    await request(koa.callback())
      .patch(`/identity/${id}/username`)
      .set("Authorization", `Bearer ${accessToken}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .send({
        username: "new-username",
      })
      .expect(204);

    await expect(TEST_IDENTITY_REPOSITORY.find({ id })).resolves.toMatchSnapshot();
  });
});
