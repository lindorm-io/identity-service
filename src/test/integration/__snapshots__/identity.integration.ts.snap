// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/identity DELETE /:id - should delete 1`] = `[RepositoryEntityNotFoundError: Entity not found]`;

exports[`/identity GET /:id - should return identity information 1`] = `
Object {
  "address": Object {
    "country": "country",
    "locality": "locality",
    "postal_code": "postalCode",
    "region": "region",
    "street_address": "streetAddress",
  },
  "birth_date": "2000-01-01",
  "display_name": "displayName#1234",
  "family_name": "familyName",
  "gender": "gender",
  "given_name": "givenName",
  "gravatar": "https://gravatar.url/",
  "locale": "sv-SE",
  "middle_name": "middleName",
  "nickname": "nickname",
  "phone_number": "+46700000000",
  "picture": "https://picture.url/",
  "preferred_username": "preferredUsername",
  "profile": "https://profile.url/",
  "username": "username",
  "website": "https://website.url/",
  "zone_info": "Europe/Stockholm",
}
`;

exports[`/identity GET /:id - should return identity information 2`] = `
Object {
  "identity": Object {
    "display_name": Object {
      "data": Array [
        Object {
          "created": 2020-01-01T07:00:00.000Z,
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "display_name_created",
              "payload": Object {
                "name": "displayName",
                "numbers": Array [
                  1234,
                ],
              },
            },
          ],
          "id": "e3926ddb-ecaf-4f66-855a-d143af54953c",
          "name": "displayName",
          "numbers": Array [
            1234,
          ],
          "updated": 2020-01-01T07:00:00.000Z,
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
    "identity": Object {
      "data": Array [
        Object {
          "address": Object {
            "country": "country",
            "locality": "locality",
            "postalCode": "postalCode",
            "region": "region",
            "streetAddress": "streetAddress",
          },
          "birthDate": "2000-01-01",
          "created": 2020-01-01T07:00:00.000Z,
          "displayName": Object {
            "name": "displayName",
            "number": 1234,
          },
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_created",
              "payload": Object {
                "address": Object {
                  "country": "country",
                  "locality": "locality",
                  "postalCode": "postalCode",
                  "region": "region",
                  "streetAddress": "streetAddress",
                },
                "birthDate": "2000-01-01",
                "familyName": "familyName",
                "gender": "gender",
                "givenName": "givenName",
                "locale": "sv-SE",
                "middleName": "middleName",
                "nickname": "nickname",
                "phoneNumber": "+46700000000",
                "picture": "https://picture.url/",
                "preferredUsername": "preferredUsername",
                "profile": "https://profile.url/",
                "website": "https://website.url/",
                "zoneInfo": "Europe/Stockholm",
              },
            },
          ],
          "familyName": "familyName",
          "gender": "gender",
          "givenName": "givenName",
          "gravatar": "https://gravatar.url/",
          "id": "e3926ddb-ecaf-4f66-855a-d143af54953c",
          "locale": "sv-SE",
          "middleName": "middleName",
          "nickname": "nickname",
          "phoneNumber": "+46700000000",
          "phoneNumberVerified": true,
          "picture": "https://picture.url/",
          "preferredUsername": "preferredUsername",
          "profile": "https://profile.url/",
          "updated": 2020-01-01T07:00:00.000Z,
          "username": "username",
          "version": 0,
          "website": "https://website.url/",
          "zoneInfo": "Europe/Stockholm",
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;

exports[`/identity PATCH /:id - should update 1`] = `
Identity {
  "_address": null,
  "_birthDate": null,
  "_created": 2020-01-01T07:00:00.000Z,
  "_displayName": null,
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "identity_created",
      "payload": Object {
        "address": null,
        "birthDate": null,
        "familyName": null,
        "gender": null,
        "givenName": null,
        "locale": null,
        "middleName": null,
        "nickname": null,
        "phoneNumber": null,
        "picture": null,
        "preferredUsername": null,
        "profile": null,
        "website": null,
        "zoneInfo": null,
      },
    },
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "identity_given_name_changed",
      "payload": Object {
        "givenName": "new-givenName",
      },
    },
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "identity_middle_name_changed",
      "payload": Object {
        "middleName": "new-middleName",
      },
    },
  ],
  "_familyName": null,
  "_gender": null,
  "_givenName": "new-givenName",
  "_gravatar": null,
  "_id": "0d573da5-97d7-44b1-8aee-9951e177d0b2",
  "_locale": null,
  "_middleName": "new-middleName",
  "_nickname": null,
  "_phoneNumber": null,
  "_phoneNumberVerified": false,
  "_picture": null,
  "_preferredUsername": null,
  "_profile": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_username": null,
  "_version": 1,
  "_website": null,
  "_zoneInfo": null,
}
`;

exports[`/identity PATCH /:id/address 1`] = `
Identity {
  "_address": Object {
    "country": "country",
    "locality": "locality",
    "postalCode": "postalCode",
    "region": "region",
    "streetAddress": "streetAddress",
  },
  "_birthDate": null,
  "_created": 2020-01-01T07:00:00.000Z,
  "_displayName": null,
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "identity_created",
      "payload": Object {
        "address": null,
        "birthDate": null,
        "familyName": null,
        "gender": null,
        "givenName": null,
        "locale": null,
        "middleName": null,
        "nickname": null,
        "phoneNumber": null,
        "picture": null,
        "preferredUsername": null,
        "profile": null,
        "website": null,
        "zoneInfo": null,
      },
    },
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "identity_address_changed",
      "payload": Object {
        "address": Object {
          "country": "country",
          "locality": "locality",
          "postalCode": "postalCode",
          "region": "region",
          "streetAddress": "streetAddress",
        },
      },
    },
  ],
  "_familyName": null,
  "_gender": null,
  "_givenName": null,
  "_gravatar": null,
  "_id": "c8909659-a490-4d6c-88d3-6a01712b7c64",
  "_locale": null,
  "_middleName": null,
  "_nickname": null,
  "_phoneNumber": null,
  "_phoneNumberVerified": false,
  "_picture": null,
  "_preferredUsername": null,
  "_profile": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_username": null,
  "_version": 1,
  "_website": null,
  "_zoneInfo": null,
}
`;

exports[`/identity PATCH /:id/username 1`] = `
Identity {
  "_address": null,
  "_birthDate": null,
  "_created": 2020-01-01T07:00:00.000Z,
  "_displayName": null,
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "identity_created",
      "payload": Object {
        "address": null,
        "birthDate": null,
        "familyName": null,
        "gender": null,
        "givenName": null,
        "locale": null,
        "middleName": null,
        "nickname": null,
        "phoneNumber": null,
        "picture": null,
        "preferredUsername": null,
        "profile": null,
        "website": null,
        "zoneInfo": null,
      },
    },
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "identity_username_changed",
      "payload": Object {
        "username": null,
      },
    },
  ],
  "_familyName": null,
  "_gender": null,
  "_givenName": null,
  "_gravatar": null,
  "_id": "9318e77b-10bf-413a-89be-ca36056b8723",
  "_locale": null,
  "_middleName": null,
  "_nickname": null,
  "_phoneNumber": null,
  "_phoneNumberVerified": false,
  "_picture": null,
  "_preferredUsername": null,
  "_profile": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_username": "new-username",
  "_version": 1,
  "_website": null,
  "_zoneInfo": null,
}
`;
